inference-engine/thirdparty/clDNN/src/network.cpp

```diff
@@ -576,19 +549,19 @@ void network_impl::execute(const std::vector<refcounted_obj_ptr<event_impl>>& ev
 
     for (auto& dout : _data_outputs) {  // data primitives are not executed so if they are marked as output we need to add
                                         // them valid events manually
-        _events[dout->id()] = get_engine().create_user_event(get_id(), true);
+        _events[dout->id()] = get_stream().create_user_event(true);
     }
 
     for (auto& prim : _primitives) {
         prim.second->reset_output_change();
     }
 
-    get_engine().get_context()->reset_events(get_id());
+    get_stream().reset_events();
 
     // Using output of previouse network as input to another one may cause hazard (in OOOQ mode) if user would not
     // provide proper event to execution. Flushing pipeline should prevent this kind of issues.
     // In scenarios with a big number of very small networks it can provide performance drop.
-    get_engine().flush_network(get_id());
+    get_stream().finish();
```

Comment:
This was needed because we had queue synchronization issues. At the current version when user wants to read output buffer, he implicitly calls `gpu_usm::lock()` method which calls `clFinish()` for **the same** queue where inference is scheduled as queue is passed to memory on object creation. After this patch, we won't require stream/queue to construct memory object anymore, and stream will be needed only to lock or copy the memory. And most of the tests use service stream (`get_test_stream()`) for output buffers lock, which means that `clFinish()` is called for **different** ocl queue, thus we could read output data before it was actually calculated. That's why initial version had `finish()` call instead of `flush()`.

In latest commits I've changed the behavior a little bit. First of all, it seems that enqueue method had invalidation condition for enabling of output events: 
`!context()->get_configuration().host_out_of_order`  which means that output event for each kernel is created for in-order queues, while the expectation is that events must be created for out-of-order queues. 
With the in-order -> out-of-order replacement in this condition, `network_output` structure has not-null event handle, and  `_event->wait();` call inside `get_memory()` method do the synchronization.

But we still have some issues with output events handling for in-order queue (no handle is created for output events in some cases), so I've added explicit finish() call for in_order queue in `get_memory()` method.

For now it's more like experimental thing, I'll need to check this approach and ensure that it doesn't have any performance/memory impacts.

Reply:
I see, thanks